File: java/src/main/java/io/dure/tree/ValidBST.java
package io.dure.tree;

import java.util.Deque;
import java.util.LinkedList;

public class ValidBST {
    public boolean isValidBST(TreeNode root) {
        return isValidBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    private boolean isValidBST(TreeNode node, long min, long max) {
        if (node == null) {
            return true;
        }
        if (node.val <= min || node.val >= max) {
            return false;
        }
        return isValidBST(node.left, min, node.val) && isValidBST(node.right, node.val, max);
    }

    public boolean isValidBST2(TreeNode root) {
        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode curr = root;
        long preVal = Long.MIN_VALUE;
        while (curr != null || !stack.isEmpty()) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
            curr = stack.pop();
            if (curr.val <= preVal) {
                return false;
            }
            preVal = curr.val;
            curr = curr.right;
        }
        return true;
    }
}

File: go/tree/validBST.go
package tree

import "math"

func isValidBST(root *TreeNode) bool {
	var dfs func(node *TreeNode, min, max int) bool
	dfs = func(node *TreeNode, min, max int) bool {
		if node == nil {
			return true
		}
		if node.Val <= min || node.Val >= max {
			return false
		}

		return dfs(node.Left, min, node.Val) && dfs(node.Right, node.Val, max)

	}
	return dfs(root, math.MinInt64, math.MaxInt64)
}

func isValidBST2(root *TreeNode) bool {
	var stack []*TreeNode
	curr := root
	preVal := math.MinInt64
	for curr != nil || len(stack) != 0 {
		for curr != nil {
			stack = append(stack, curr)
			curr = curr.Left
		}
		curr = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if curr.Val <= preVal {
			return false
		}
		preVal = curr.Val
		curr = curr.Right
	}
	return true
}
